File: LeetCodePractice/src/BFS/WordLadder.java
package BFS;
/*
Given two words beginWord and endWord, and a dictionary wordList, return the length of the shortest transformation sequence from beginWord to endWord, such that:

Only one letter can be changed at a time.
Each transformed word must exist in the word list.
Return 0 if there is no such transformation sequence.



Example 1:

Input: beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log","cog"]
Output: 5
Explanation: As one shortest transformation is "hit" -> "hot" -> "dot" -> "dog" -> "cog", return its length 5.
Example 2:

Input: beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log"]
Output: 0
Explanation: The endWord "cog" is not in wordList, therefore no possible transformation.


Constraints:

1 <= beginWord.length <= 100
endWord.length == beginWord.length
1 <= wordList.length <= 5000
wordList[i].length == beginWord.length
beginWord, endWord, and wordList[i] consist of lowercase English letters.
beginWord != endWord
All the strings in wordList are unique.
*/

import java.util.*;

public class WordLadder {
	/**
     * @param start, a string
     * @param end, a string
     * @param wordList, a list of string
     * @return an integer
     */
   public int ladderLength(String start, String end, List<String> wordList) {
       Set<String> dict = new HashSet<>(wordList);
       int dist = 1;
       HashSet<String> visited = new HashSet<>();
       Queue<String> queue = new LinkedList<>();
       queue.offer(start);
       visited.add(start);
       while(!queue.isEmpty()){
           int size = queue.size();
           for(int i = 0; i < size; i++){
               String cur = queue.poll();
               if(end.equals(cur)){
                   return dist;
               }
               List<String> neighbors = findNeighbors(cur, dict);
               for(String nb : neighbors){
                   if(!visited.contains(nb)){
                       queue.offer(nb);
                       visited.add(nb);
                   }
               }
           }
           dist++;
       }
       return 0;
   }
   private List<String> findNeighbors(String str, Set<String> dict){
       List<String> res = new ArrayList<>();
       int len = str.length();
       for(int i = 0; i < len; i++){
           for(char c = 'a'; c <= 'z'; c++){
               if(c != str.charAt(i)){
                   String temp = replace(str, i, c);
                   if(dict.contains(temp)){
                       res.add(temp);
                   }
               }
           }
       }
       return res;
   }
   
   private String replace(String s, int i, char c){
       char[] temp = s.toCharArray();
       temp[i] = c;
       return new String(temp);
   }
}

File: LeetCodePython/WordLadder.py
"""
A transformation sequence from word beginWord to word endWord using a dictionary wordList is a sequence of words beginWord -> s1 -> s2 -> ... -> sk such that:

Every adjacent pair of words differs by a single letter.
Every si for 1 <= i <= k is in wordList. Note that beginWord does not need to be in wordList.
sk == endWord
Given two words, beginWord and endWord, and a dictionary wordList, return the number of words in the shortest transformation sequence from beginWord to endWord, or 0 if no such sequence exists.



Example 1:
Input: beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log","cog"]
Output: 5
Explanation: One shortest transformation sequence is "hit" -> "hot" -> "dot" -> "dog" -> cog", which is 5 words long.

Example 2:
Input: beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log"]
Output: 0
Explanation: The endWord "cog" is not in wordList, therefore there is no valid transformation sequence.

Constraints:
1 <= beginWord.length <= 10
endWord.length == beginWord.length
1 <= wordList.length <= 5000
wordList[i].length == beginWord.length
beginWord, endWord, and wordList[i] consist of lowercase English letters.
beginWord != endWord
All the words in wordList are unique.

anlaysis:
BFS
TC: O(M^2 * N) M is the word character length, N is th wordlist length
"""
import collections
from typing import List


class WordLadder:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        wordList = set(wordList)
        queue = collections.deque([(beginWord, 1)])
        while queue:
            w, l = queue.popleft()
            if w == endWord:
                return l
            for i in range(len(w)):
                for c in 'abcdefghijklmnopqrstuvwxyz':
                    nb = w[:i] + c + w[i + 1:]
                    if nb in wordList:
                        wordList.remove(nb)
                        queue.append((nb, l + 1))
        return 0

    def ladderLengthBidirectionBFS(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        wordList = set(wordList)
        if endWord not in wordList:
            return 0
        visited = set()
        begin_set = {beginWord}
        end_set = {endWord}
        step = 1
        while begin_set and end_set:
            next_set = set()
            for w in begin_set:
                if w in end_set:
                    return step
                visited.add(w)
                for i in range(len(w)):
                    for c in 'abcdefghijklmnopqrstuvwxyz':
                        nb = w[:i] + c + w[i + 1:]
                        if nb in wordList and nb not in visited:
                            next_set.add(nb)
            if len(end_set) < len(next_set):
                begin_set = end_set
                end_set = next_set
            else:
                begin_set = next_set
            step += 1
        return 0
